src/axolotl/prompt_strategies/chat_template.py

```diff
@@ -79,62 +201,169 @@ def messages(self, messages):
+
+        Args:
+            conversation_ids (list[int]): Token IDs representing the conversation.
+            turn (int): The turn number to locate (based on EOS tokens).
+            turn_content (str): String containing the content of the turn.
+
+        Returns:
+            tuple: (start_idx, end_idx) indices of the start and end of the turn content.
+                   Returns (-1, -1) if the turn content is not found.
+        """
+        content = turn_content.get(self.prompter.message_field_content, "")
+        content_ids = self.tokenizer.encode(content, add_special_tokens=False)
+
+        eos_token_id = self.tokenizer.eos_token_id
+        eos_count = 0
+        start_search_idx = 0
+
+        # Locate the starting index after the specified number of EOS tokens
+        for i, token_id in enumerate(conversation_ids):
+            if token_id == eos_token_id:
+                eos_count += 1
+                if eos_count == turn:
+                    start_search_idx = (
+                        i + 1
+                    )  # Start searching after the specified turn's EOS token
+                    break
+
+        # Find the start index of the content within the conversation
+        start_idx = -1
+        for i in range(start_search_idx, len(conversation_ids) - len(content_ids) + 1):
+            if conversation_ids[i : i + len(content_ids)] == content_ids:
+                start_idx = i
+                break
+
+        if start_idx != -1:
+            end_idx = start_idx + len(content_ids)
+        else:
+            end_idx = -1
+
+        return start_idx, end_idx
 
     def get_conversation_thread(self, prompt):
         return prompt[self.messages]
 
 
 def load(tokenizer, cfg, ds_cfg: Optional[Dict[str, Any]] = None):
-    chat_template = (
-        ds_cfg["chat_template"] if ds_cfg and "chat_template" in ds_cfg else "chatml"
-    )
-    message_field_role = (
-        ds_cfg["message_field_role"]
-        if ds_cfg and "message_field_role" in ds_cfg
-        else "from"
-    )
-    message_field_content = (
-        ds_cfg["message_field_content"]
-        if ds_cfg and "message_field_content" in ds_cfg
-        else "value"
-    )
-    roles = ds_cfg["roles"] if ds_cfg and "roles" in ds_cfg else None
-    drop_system_message = (
-        ds_cfg["drop_system_message"]
-        if ds_cfg and "drop_system_message" in ds_cfg
-        else False
-    )
+    ds_cfg = ds_cfg or {}
 
-    strategy = ChatTemplateStrategy(
-        ChatTemplatePrompter(
-            tokenizer,
-            chat_templates(chat_template),
-            message_field_role=message_field_role,
-            message_field_content=message_field_content,
-            roles=roles,
-            drop_system_message=drop_system_message,
+    prompter_params = {
+        "tokenizer": tokenizer,
+        "chat_template": chat_templates(ds_cfg.get("chat_template", "chatml")),
+        "message_field_role": ds_cfg.get("message_field_role", "from"),
```

Comment:
```suggestion
        "max_length": cfg.sequence_len,,
        "message_field_role": ds_cfg.get("message_field_role", "from"),
```
I realized I had originally forgotten to include this. without it, everything gets truncated to the default of 2k.

Reply:
So there has already been some confusion about this, because there was already the additional `max_length` config option added (which as you said, apparently wasn't passed in).

Should we just be using global sequence_len throughout, and get rid of the dataset configuration for max_length?

Reply:
This is really to control the length of the outputs of the tokenizer. Here's something I have in mind that might be more flexible and give users more control over whether samples should be dropped or truncated if they are too long.

<img width="762" alt="Screenshot 2024-07-23 at 10 48 58 PM" src="https://github.com/user-attachments/assets/8c4205cb-d16c-493a-b545-b39cef9e3892">